test cases/common/251 subproject dependency variables/subprojects/subfiles/meson.build

```diff
@@ -0,0 +1,26 @@
+project('dependency variable resource')
+
+files_dep = declare_dependency(
+    variables: [
+      'pkgdatadir=@0@/subdir'.format(meson.current_source_dir()),
+      'pkgdatadir2=@0@/subdir2'.format(meson.current_source_dir()),
+    ]
+)
```

Comment:
This is ... less than ideal.This is doing a complete end run around sandboxing and further the master project needs to know how the subproject has arranged its files. By corollary if you update the subproject so that it changes its file locations (which is its implementation detail) then things will fail to configure. The proper syntax for this is the regular:

```

files1_dep = declare_dependency(
    sources: files('subdir/foo.c'
)
```

This already works today, is less code for end users to write and fully obeys sandboxing.

Reply:
No? It's not an end run around sandboxing at all. The master project needs to know *nothing* about how the subproject arranges its files. Rather, the master project needs to know that the subproject has documented:
- use `pkg-config --variable pkgdatadir` to retrieve a directory such as /usr/share/subdir/
- this installed directory has important data files, namely foo.xml or foo.c

So, the master project looks up a dependency, and asks that dependency where the installed datadir is. However, subprojects may specify that variable too, in order to provide the same *interface* as the installed project.

In order to pedantically guarantee that no one is *allowed* to violate the sandbox, we ensure two things:
- you are not permitted to use .. in this workflow
- the only files you are allowed to retrieve from the sandbox are ones that are marked for public installation

You cannot use a declare_dependency "sources" for data files like XML files to be processed by the wayland-protocols scanner, so that has never worked and never will work. My test case doesn't depend on wayland-scanner, so I figured using executable() and source files was the simplest way of testing the sandbox handling.

Also, even if this were only about getting source files for an executable, that declare_dependency requires special casing the difference between a subproject and a system dependency. In real use cases, you don't do `dependency('subfiles')` and expect an *installed* pkg-config dependency to provide source files as cflags. So you cannot use that declare_dependency as a wrap `[provide]` fallback for a system dependency.

Reply:
> You cannot use a declare_dependency "sources" for data files like XML files to be processed by the wayland-protocols scanner, so that has never worked and never will work.

Then that is the actual feature we need to provide, a way for declared dependencies to carry arbitrary extra files. Maybe something like:

```
declare_dependency(extra_files: files('bob.xml'))
```

Alternatively there has been talk of adding a `directories` object in addition to a `files` one for cases like these. In that case you could more properly validate that when it is used in a different subproject you are not allowed to add any `..` segments to it (possibly any directory path segments).

Also based on a quick reading I already found a potential way to use this to escape the sandbox (I did not test if it actually works):

```
d = declare_dependency(variables: ['nastydir=path/to/a/different/subproject/'])
h4cker_file = d.get_variable('nastydir') / 'internal_file.c'
```

In any case storing information like this in plain strings is *bad* and must *never* be done. The innards should be hidden in an opaque object. Have we learned nothing from CMake?

In addition, new code should not use `join_paths`. We should probably deprecate that if we have not already.

Reply:
> declare_dependency(extra_files: files('bob.xml'))

This sounds useful, but how would one make use of this? Do we need to add `dep.get_extra_file()` as well, which would look at pkgconfig variables or internal `extra_files`? We probably want to make `extra_files` a dict instead of a list?

One last thing which would be useful is to allow targets in `extra_files` as well, to allow executables and generated files to be passed from a subproject to its parent project.

Reply:
> Then that is the actual feature we need to provide, a way for declared dependencies to carry arbitrary extra files. Maybe something like:
> 
> ```
> declare_dependency(extra_files: files('bob.xml'))
> ```

Please suggest how this can work with a system installed pkg-config dependency, if it doesn't work with a system-installed pkg-config dependency then projects will, rightfully, refuse to use it because they want to write code that works with system-installed dependencies but have subproject fallbacks "do the right thing".

> Alternatively there has been talk of adding a `directories` object in addition to a `files` one for cases like these. In that case you could more properly validate that when it is used in a different subproject you are not allowed to add any `..` segments to it (possibly any directory path segments).

This is LITERALLY a directories object, except that it doesn't require:
- adding a new function
- editing all existing meson.build which currently provides `declare_dependency(variables: ...)` as strings, to provide `variables: dir('.')` instead
- forcing all projects that want to make use of this functionality to bump the minimum version of Meson to the very latest bleeding edge version of Meson just so they can describe a string, which is used as a string and joined as a string, via a special function that returns `class DependencyVariableString(str)` except sed'ed out to `class DirectoryString(str)`, at which point we are exactly back to where I was in this PR, but with one more meson function to document.

> Also based on a quick reading I already found a potential way to use this to escape the sandbox (I did not test if it actually works):

It will not work, unless `path/to/a/different/subproject/` declares `internal_file.c` as an `install_data()` in which case I guess it is not actually an internal file. I considered this possibility and guarded against it.

In fact, my successfully figuring out how to ban this so that it doesn't work, is what made me confident enough to take this PR out of draft mode and consider it a viable PR.

> In any case storing information like this in plain strings is _bad_ and must _never_ be done. The innards should be hidden in an opaque object. Have we learned nothing from CMake?

Again, the ground source of truth is `/usr/share/wayland-protocols/stable/xdg-shell/xdg-shell.xml` and that *must* be handled as a string, and passed as a string as `input: ...` to a custom_target.

This PR allows `dependency('wayland-protocols').get_variable('pkgdatadir') / 'stable/xdg-shell/xdg-shell.xml'`, which successfully works ***and is the correct source specification methodology*** if get_variable() evaluates to /usr/share/wayland-protocols, to *also* work as a subproject where get_variable() evaluates to some install_data or install_subdir provided by the subproject.

> In addition, new code should not use `join_paths`. We should probably deprecate that if we have not already.

This is the wrong PR to discuss deprecating join_paths, and since this PR is backwards-compatible to meson.build files specifying a minimum version of meson from before any proposed deprecation, this PR *must* work with `join_paths()`.

Reply:
> to allow targets in `extra_files` as well, to allow executables and generated files to be passed from a subproject to its parent project.

No, if you want a program installed on the system you want to use find_program() and we already handle that via program overrides.

And per my consistent theme in this discussion, I want subprojects to transparently act like system installed dependencies, which means we cannot have people doing if/else and using find_program if dep.type_name() evaluates to pkgconfig, but doing some declare_dependency thing if it evaluates to 'internal'.

The point of a subproject is to be transparent, so the code originally written for a system dependency should, as a Meson development goal, continue to work without modification if a dependency is of type 'internal'.

That's all I want here. This PR is in aid of that. No new API.

Reply:
Eli, I actually agree with Jussi, as you know from our discussion on Matrix.

> This is LITERALLY a directories object, except that it doesn't require:
>  *  adding a new function

I think when we find ourselves painted into a corner because of our older API decisions, adding new API based on our learnings from that is a good thing.

>  *  editing all existing meson.build which currently provides declare_dependency(variables: ...) as strings, to provide variables: dir('.') instead

We've had to do this in the past a lot for subprojects (with gstreamer and GNOME), and we did it anyway because it led us to the right API (`[provide]` sections in wraps).

>  *  forcing all projects that want to make use of this functionality to bump the minimum version of Meson to the very latest bleeding edge version of Meson

They don't have to require the latest version, the code can be conditional on the current meson version. You need the latest meson version to get things to work anyway.

>  just so they can describe a string, which is used as a string and joined as a string, 

Except that it's not a string. It's a path. Strings are not paths. If someone is using strings as paths, that's because there are no APIs for path manipulation. A good example is `os.path` vs `pathlib`. I don't think we should continue down that path. I said the same thing on Matrix, and you seemed to agree (edited for clarity):

```
<nirbheek> I was just thinking that anytime we're using a string for a path, it's because we don't have the correct abstraction in place
<nirbheek> And if we're adding an abstraction here, we might as well make it the right one?
...
<nirbheek> I just think that `DependencyVariableString` is the wrong abstraction? Maybe we should have `dirs()` and the `includedir` is a `dirs()` for `declare_dependency` and you can get `files()` in the subdirs of that dir with `/`
<nirbheek> i.e., make the object that is in `variables:` rich instead of making the internal implementation of `variables:` rich?
<elibrokeit> so you want `declare_dependency` to provide some object type that isn't a string?
<elibrokeit> like `declare_dependency(variables: {'foodir': dirs('.')})`
<nirbheek> Yes
<elibrokeit> adding a `dirs()` function to the meson API for this seems a bit too niche, and I guess it would need to return a dedicated meson object type that subclasses a stringholder anyway
<nirbheek> I don't know if it's niche -- I think this is just the first compelling use-case for it
```

> via a special function that returns class DependencyVariableString(str) except sed'ed out to class DirectoryString(str), at which point we are exactly back to where I was in this PR, but with one more meson function to document.

Not quite. I suggested `dirs()` above, and so did Jussi, and it would work differently than what you say. On further thinking, I think `path()` makes more sense.

Producer:
```
proj_dep = declare_dependency(...
    # We should eventually deprecate the 'key=value' syntax, it's too stringy
    variables: {'datadir': path('.')})
```

Consumer:
```
proj_dep = dependency('proj-name')
datadir = proj_dep.get_variable('datadir')
protocol_xml = datadir / 'protocol.xml'
```

When `proj_dep` is provided by a system dep (pkgconfig, cmake, etc), `datadir` would be a string, because we have no way to know what the variable actually is.

When `proj_dep` is provided by a subproject, `datadir` would be a `path()` object, which can also be used as a string implicitly.

In both cases, `protocol_xml` should be a `path()` object, i.e., using the `/` operator on a `str` or a `path()` object should yield a `path()` object (which is usable interchangeably as an `str`).

We can either have a `path()` object be converted to a `files()` object explicitly with `.to_files()`, or implicitly by having all API that accepts `files()` also accept `path()`. I am not sure which one is better here, but usually explicit is better.

Reply:
I apologize if my attempt to gently cast doubt on your proposal by pointing out that it adds niche API which has no practical use, sounded like I was agreeing with you.

That was never my intention. I continue to maintain that this entire line of thought is 100% wrong.

I simply cannot fathom your objection here. And I once more want to reiterate, that you're saying "not quite" and implying that `dirs()` or `path()` is somehow different from `DependencyVariableString(str)` while in fact they are the same thing, because as you pointed out they can be used interchangeably.

I would like projects to just work the expected way, and have meson intelligently detect when projects are doing the right thing and penalizing them for that right thing involving a subproject. You want to add new API that forces every project to edit their build files, adding additional cognitive overhead to the build language, and the ***only*** thing it does is produce a string which has been marked at the time of declaration as "not a sandbox violation" rather than at time of variable retrieval. Yes, that's right. You already explained that a path object would in all cases be usable as a string, and of course strings must always be usable as paths so that we can join them and use them as arguments to functions that expect strings describing a path (and which years of existing build files in thousands of projects are reasonably and sensibly using).

The actual defense against sandbox violations is my check that you can only use installed data files, and you need that check no matter what niche and cognitively strange API you propose I add to declare_dependency.

I am saddened that my attempt to make things better for Wayland is not going to be accepted, but oh well. Life isn't perfect and even though meson can't sensibly use subprojects in this case, it's still a better build system than cmake.

Reply:
>  dirs() or path() is somehow different from DependencyVariableString(str) while in fact they are the same thing,

There is a major semantic difference. The return value of dirs could only point to an existing directory on the file system. It would be verified on creation and thus would be safer to use. A `DependencyVariableString` has no such guarantee, it would need to be validated on use.

Another thing is that Directory (or Path) objects could be extended to relative paths in command invocations. DependencyVariableStrings would have to be absolute. This is actually important for deterministic builds. I had to change the implementation a fair bit ages ago to use relative paths whenever possible because absolute paths do not work (I do not know why, just that this is what I had to to on request by the reproducible builds folks).

Reply:
Anyway, I implemented this check anyway, even though it is pointless.

I also closed the door on
```meson
d = declare_dependency(variables: ['nastydir=path/to/a/different/subproject/'])
```

Note: ***even though*** this cannot ever work unless it refers to installed files.

However, with the latest push, now it won't work even to poke at another project's installed files, unless that project does the declare_dependency itself. Because it is tracked during declare_dependency, so we actually know which subproject it is part of.